include/revng/Model/Processing.h

```diff
@@ -12,19 +12,44 @@ namespace model {
 
 /// Given \p Types, drop all the types and DynamicFunctions depending on it
 ///
-/// Sometimes you createe a set of placeholder types in the model, but they end
+/// Sometimes you create a set of placeholder types in the model, but they end
 /// up being invalid. In that case, they, and all the types depending on them,
 /// need to be dropped.
 ///
 /// \return the number of dropped types
 unsigned dropTypesDependingOnTypes(TupleTree<model::Binary> &Binary,
                                    const std::set<const model::Type *> &Types);
 
+/// A specialization of dropTypesDependingOnTypes() that additionally finds
+/// the types to remove based on whether they are referenced by any other types
+unsigned dropUnreferencedTypes(TupleTree<model::Binary> &Binary);
+
+/// An overload allowing to limit the unreferenced type removal to a specific
+/// type kind (for example, structs only, primitives only, etc).
+unsigned dropUnreferencedTypes(TupleTree<model::Binary> &Binary,
+                               model::TypeKind::Values Kind);
```

Comment:
Mostly `while (dropUnreferencedTypes(...)) ;` is just a quick fix. Since I wanted to push the pr in faster. I just ran into the bug for the fact that `dropUnreferencedTypes(...)` is not the same as `dropUnreferencedTypes(...); dropUnreferencedTypes(...)`. And fixing that would mean rewriting the whole function from scratch (to keep track of the types that are only depended on by other types they would get removed).

Reply:
I was suggesting to rename `dropUnreferencedTypes` to `dropUnreferencedTypesImpl` and have `dropUnreferencedTypes` do `while (dropUnreferencedTypesImpl) ;`.

Anyway, I'd like to better understand the use case here.
For different purposes I'm implementing something to purge useless types.
My current plan is to write a function that:
1. Collects unreferenced types **or** SCC of nodes that have no references outside the SCC itself.
2. If any type in the set of nodes has a name, bail out.
3. Drop all the types.

Would that work for you? Is there some extension I can make to serve your use case too?

Reply:
Yes, the centralized solution is everything I could dream of! The only worry I have is that it's going to remove some stuff that's not really related to what I'm doing - but any simple-ish solution would do the same - as such if it's capable of removing unreferenced structs reliably - it's perfect for what I need.

Reply:
And, on top of that, if we every move to "nested model" return style, even that would stop being relevant. If I don't have to add structs - I don't need to remove them either.